js/state/syncUtil.js

```diff
@@ -87,22 +83,64 @@ const applySiteRecord = (record) => {
     const parentFolderObjectId = siteProps.parentFolderObjectId
     if (parentFolderObjectId && parentFolderObjectId.length > 0) {
       siteProps.parentFolderId =
-        getFolderIdByObjectId(new Immutable.List(parentFolderObjectId))
+        getFolderIdByObjectId(new Immutable.List(parentFolderObjectId), appState)
     }
   }
   const siteDetail = new Immutable.Map(pickFields(siteProps, SITE_FIELDS))
+  return {siteDetail, tag, existingObjectData}
+}
 
-  switch (record.action) {
-    case writeActions.CREATE:
-      appActions.addSite(siteDetail, tag, undefined, undefined, true)
-      break
-    case writeActions.UPDATE:
-      appActions.addSite(siteDetail, tag, existingObjectData, null, true)
-      break
-    case writeActions.DELETE:
-      appActions.removeSite(siteDetail, tag, true)
-      break
+/**
+ * Apply bookmark or historySite SyncRecords to the browser data store.
+ * @param {Array.Object} records
+ * @param {Immutable.Map} siteDetail
+ */
+const applySiteRecords = (records) => {
+  const appActions = require('../actions/appActions')
+  const AppStore = require('../stores/appStore')
+
+  const batchedCreates = {} // map of tag to data
+
+  const appState = AppStore.getState()
+  records.forEach((record) => {
+    if (record.action !== writeActions.CREATE) {
+      return
+    }
+    const siteData = getSiteDataFromRecord(record, appState)
+    const siteDetail = siteData.siteDetail
+    const key = siteData.tag || 'history'
+    if (batchedCreates[key]) {
+      batchedCreates[key].push(siteDetail)
+    } else {
+      batchedCreates[key] = [siteDetail]
+    }
+  })
+
+  // Create all bookmarks in a batch
+  for (var tag in batchedCreates) {
+    appActions.addSite(new Immutable.List(batchedCreates[tag]),
+      tag === 'history' ? undefined : tag,
+      undefined, undefined, true)
   }
+
+  // Apply updates and deletes after creation finishes
+  setTimeout(() => {
+    const appState = AppStore.getState()
+    records.forEach((record) => {
+      if (record.action === writeActions.CREATE) {
+        return
+      }
+      const siteData = getSiteDataFromRecord(record, appState)
+      const siteDetail = siteData.siteDetail
+      const tag = siteData.tag
+      const existingObjectData = siteData.existingObjectData
+      if (record.action === writeActions.UPDATE) {
+        appActions.addSite(siteDetail, tag, existingObjectData, null, true)
+      } else if (record.action === writeActions.DELETE) {
+        appActions.removeSite(siteDetail, tag, true)
+      }
+    })
+  }, 2000)
```

Comment:
if this has to happen after creation a timeout doesn't seem very safe, or is the timeout just an extra buffer? If so I don't think it's really necessary with the dispatcher changes